src/envoy/tcp/metadata_exchange/metadata_exchange.cc

```diff
@@ -224,31 +238,70 @@ void MetadataExchangeFilter::tryReadProxyData(Buffer::Instance& data) {
-    setMetadata(UpstreamDynamicDataKey, struct_metadata);
-  } else {
-    setMetadata(DownstreamDynamicDataKey, struct_metadata);
+  const auto key_metadata_it =
+      value_struct.fields().find(ExchangeMetadataHeader);
+  if (key_metadata_it != value_struct.fields().end()) {
+    setMetadata(config_->filter_direction_ == FilterDirection::Downstream
+                    ? UpstreamMetadataKey
+                    : DownstreamMetadataKey,
+                key_metadata_it->second);
+  }
+  const auto key_metadata_id_it =
+      value_struct.fields().find(ExchangeMetadataHeaderId);
+  if (key_metadata_id_it != value_struct.fields().end()) {
+    setMetadata(config_->filter_direction_ == FilterDirection::Downstream
+                    ? UpstreamMetadataIdKey
+                    : DownstreamMetadataIdKey,
+                key_metadata_id_it->second);
   }
 }
 
-void MetadataExchangeFilter::setMetadata(const std::string key,
-                                         const ProtobufWkt::Struct& value) {
-  read_callbacks_->connection().streamInfo().setDynamicMetadata(key, value);
+void MetadataExchangeFilter::setMetadata(
+    const std::string& key, const Envoy::ProtobufWkt::Value& value) {
+  read_callbacks_->connection().streamInfo().filterState().setData(
+      key, std::make_unique<MetadataExchangeState>(value),
+      StreamInfo::FilterState::StateType::Mutable);
 }
 
-std::unique_ptr<const google::protobuf::Struct>
-MetadataExchangeFilter::getMetadata(const std::string& key) {
+void MetadataExchangeFilter::setMetadataStringValue(
+    const std::string& key, const std::string& str_value) {
+  Envoy::ProtobufWkt::Value value;
+  value.set_string_value(str_value);
+  setMetadata(key, value);
+}
+
+void MetadataExchangeFilter::getMetadata(const std::string& key,
+                                         google::protobuf::Struct* metadata) {
   if (local_info_.node().has_metadata()) {
-    auto metadata_fields = local_info_.node().metadata().fields();
-    auto node_metadata = metadata_fields.find(key);
-    if (node_metadata != metadata_fields.end()) {
-      return std::make_unique<const google::protobuf::Struct>(
-          node_metadata->second.struct_value());
+    google::protobuf::Struct node_metadata = local_info_.node().metadata();
+    const auto key_it = node_metadata.fields().find(key);
+    if (key_it == node_metadata.fields().end()) {
+      return;
+    }
+
+    const auto& keys_value = key_it->second;
+    if (keys_value.kind_case() != google::protobuf::Value::kStringValue) {
+      return;
+    }
+
+    // select keys from the metadata using the keys
+    const std::set<std::string> keys =
+        absl::StrSplit(keys_value.string_value(), ',', absl::SkipWhitespace());
+    for (auto key : keys) {
+      const auto entry_it = node_metadata.fields().find(key);
+      if (entry_it == node_metadata.fields().end()) {
+        continue;
+      }
+      (*metadata->mutable_fields())[key] = entry_it->second;
     }
```

Comment:
@kyessenov : a lot of this code is copied from your http metadata exchange filter.. should we add it later to a common util?

Reply:
Please do. `extensions/common` seems like the appropriate place for this function.

Reply:
Done!